packages/daemon/DESIGN.md

```diff
@@ -0,0 +1,90 @@
+
+Endo is a user agent for running guest programs.
+These programs receive capabilities from the user in the form of references to
+potentially powerful remote objects.
+To that end, Endo exhibits a system called Eventual Send that stretches
+references to remote objects over one or more capability transfer protocols
+(CapTP).
+
+A Promise represents a result from either the past or the future, and Eventual
+Send enables programs to interact with those eventual results, even if they are
+in another worker or on the far side of a network.
+Furthermore, Eventual Send allows messages to be pipelined to these eventual
+results to overcome any latency that might otherwise accrue if the program
+waited for each intermediate result before sending a follow-up message.
+
+Much hinges on what happens when a program restarts.
+Like most programs, Endo workers start without any memory from their previous
+run.
+
+Endo applications ultimately receive all their powers from the user.
+They receive a "powerbox" object from the user they can use to petition the
+user for access to additional references including capability-bearing objects,
+including remote objects, including remote objects from other peers.
+If the program requests a reference with a name, Endo records a formula for
+recreating the reference such that any subsequent request with the same
+name will provide the result of applying the same formula without interrogating
+the user.
+
+Likewise, the Endo user holds their own powerbox and can assign pet names to
```

Comment:
Ok, I am genuinely confused at this point. The program receives a powerbox from the user, which means that the user in some sense had that powerbox. But this is a distinct powerbox held by that same user?

Reply:
I’ve posted an update that clarifies how programs request powers, how users review requests, and how both have pet name spaces.